builtin/provisioners/puppet/resource_provisioner.go

```diff
@@ -0,0 +1,332 @@
+				Optional: true,
+			},
+			"environment": &schema.Schema{
+				Type:     schema.TypeString,
+				Default:  "production",
+				Optional: true,
+			},
+			"bolt_timeout": &schema.Schema{
+				Type:     schema.TypeString,
+				Default:  "5m",
+				Optional: true,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					_, err := time.ParseDuration(val.(string))
+					if err != nil {
+						errs = append(errs, err)
+					}
+					return warns, errs
+				},
+			},
+		},
+		ApplyFunc: applyFn,
+	}
+}
+
+func applyFn(ctx context.Context) error {
+	output := ctx.Value(schema.ProvOutputKey).(terraform.UIOutput)
+	state := ctx.Value(schema.ProvRawStateKey).(*terraform.InstanceState)
+	configData := ctx.Value(schema.ProvConfigDataKey).(*schema.ResourceData)
+
+	p, err := decodeConfig(configData)
+	if err != nil {
+		return err
+	}
+
+	p.instanceState = state
+	p.output = output
+
+	if p.OSType == "" {
+		switch connType := state.Ephemeral.ConnInfo["type"]; connType {
+		case "ssh", "":
+			p.OSType = "linux"
+		case "winrm":
+			p.OSType = "windows"
+		default:
+			return fmt.Errorf("Unsupported connection type: %s", connType)
+		}
+	}
+
+	switch p.OSType {
+	case "linux":
+		p.runPuppetAgent = p.linuxRunPuppetAgent
+		p.installPuppetAgent = p.linuxInstallPuppetAgent
+		p.uploadFile = p.linuxUploadFile
+		p.defaultCertname = p.linuxDefaultCertname
+	case "windows":
+		p.runPuppetAgent = p.windowsRunPuppetAgent
+		p.installPuppetAgent = p.windowsInstallPuppetAgent
+		p.uploadFile = p.windowsUploadFile
+		p.UseSudo = false
+		p.defaultCertname = p.windowsDefaultCertname
+	default:
+		return fmt.Errorf("Unsupported OS type: %s", p.OSType)
+	}
+
+	comm, err := communicator.New(state)
+	if err != nil {
+		return err
+	}
+
+	retryCtx, cancel := context.WithTimeout(ctx, comm.Timeout())
+	defer cancel()
+
+	err = communicator.Retry(retryCtx, func() error {
+		return comm.Connect(output)
+	})
+	if err != nil {
+		return err
+	}
+	defer comm.Disconnect()
+
+	p.comm = comm
+
+	if p.OpenSource {
+		p.installPuppetAgent = p.installPuppetAgentOpenSource
+	}
+
+	csrAttrs := new(csrAttributes)
+	csrAttrs.CustomAttributes = make(map[string]string)
+	for k, v := range p.CustomAttributes {
+		csrAttrs.CustomAttributes[k] = v.(string)
+	}
+
+	csrAttrs.ExtensionRequests = make(map[string]string)
+	for k, v := range p.ExtensionRequests {
+		csrAttrs.ExtensionRequests[k] = v.(string)
+	}
+
+	if p.Autosign {
+		if p.Certname == "" {
+			p.Certname, _ = p.defaultCertname()
+		}
+
+		autosignToken, err := p.generateAutosignToken(p.Certname)
+		if err != nil {
+			return fmt.Errorf("Failed to generate an autosign token: %s", err)
+		}
+		csrAttrs.CustomAttributes["challengePassword"] = autosignToken
+	}
+
+	if err = p.writeCSRAttributes(csrAttrs); err != nil {
+		return fmt.Errorf("Failed to write csr_attributes.yaml: %s", err)
+	}
+
+	if err = p.installPuppetAgent(); err != nil {
+		return err
+	}
+
+	if err = p.runPuppetAgent(); err != nil {
```

Comment:
I imagine documentation will cover this, but what's the expected workflow if autosign is false? Would a user just expect the puppet run fail until the cert is signed?

Reply:
If autosign is false, the Puppet run will probably fail. If they've configured their master to use naive certname matching autosigning (i.e. just trusting and signing any request for `*.mycompany.com`) then that will succeed. For other cases, I'm assuming they'll use something like a custom remote-exec provisioner before the puppet provisioner to take whatever action their setup requires (adding an entry to their CMBD perhaps).